monkey/monkey_island/cc/repositories/i_otp_repository.py

```diff
@@ -0,0 +1,44 @@
+from abc import ABC, abstractmethod
+
+from monkey_island.cc.models import OTP
+
+
+class IOTPRepository(ABC):
+    """A repository used to store and retrieve `OTP`s"""
+
+    @abstractmethod
+    def save_otp(self, otp: OTP):
+        """
+        Insert an OTP into the repository
+
+        :param otp: The OTP to insert
+        :raises StorageError: If an error occurs while attempting to insert the OTP
+        """
+
+    @abstractmethod
+    def get_otp(self, otp: str) -> OTP:
+        """
+        Get an OTP from the repository
+
+        :param otp: The ID of the OTP to get
+        :return: The OTP
+        :raises RetrievalError: If an error occurs while attempting to retrieve the OTP
+        :raises UnknownRecordError: If the OTP was not found
+        """
+
+    @abstractmethod
+    def delete_otp(self, otp: str):
+        """
+        Delete an OTP from the repository
+
+        :param otp: The OTP to delete
+        :raises RemovalError: If an error occurs while attempting to delete the OTP
+        """
```

Comment:
I think we decided not to allow deletion. One-time passwords should be one-time. So leaving them in the database allows us to ensure we don't reissue the same OTP for the life of the Island.

On that note, perhaps `reset()` should be `invalidate_all_otps()` and simply set the expiration times to sometime in the past (i.e. t=0.0).

Reply:
We still want to mark an OTP as "expired" so that it can no longer be used to authenticate. Would `expire_otp(otp, time)` suffice?

Reply:
What if we just insert the time in the database when the token will be expired? Then we only ever need to check if `time.monotonic() > expiration_time`. We can do that when the OTP is first inserted into the database.